hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/HoodieCompactor.java

```diff
@@ -130,9 +159,27 @@ public HoodieData<WriteStatus> compact(
     TaskContextSupplier taskContextSupplier = table.getTaskContextSupplier();
     // if this is a MDT, set up the instant range of log reader just like regular MDT snapshot reader.
     Option<InstantRange> instantRange = CompactHelpers.getInstance().getInstantRange(metaClient);
-    return context.parallelize(operations).map(operation -> compact(
-        compactionHandler, metaClient, config, operation, compactionInstantTime, maxInstantTime, instantRange, taskContextSupplier, executionHelper))
-        .flatMap(List::iterator);
+
+    // do one upfront processing to detect if we need partial update.
+    // if yes, call prepareBroadcast.
+    HoodiePairData<CompactionOperation, Boolean> operationPartialUpdatePair = context.parallelize(operations)
+        .mapToPair(operation -> Pair.of(operation, containsPartialUpdate(compactionHandler, metaClient, config, operation, compactionInstantTime, maxInstantTime,
+            instantRange, taskContextSupplier, executionHelper)));
+    // should we persist operationPartialUpdatePair?
+
+    boolean containsPartialUpdate = operationPartialUpdatePair.values().collectAsList().stream().anyMatch(partialUpdate -> partialUpdate);
+    Option<PartialUpdateReaderContext> partialUpdateReaderContextOpt = Option.empty();
+    if (containsPartialUpdate) {
+      partialUpdateReaderContextOpt = getPartialUpdateReaderContext(context);
+      // Broadcast ParquetFileReader for file group reader if needed.
+      ValidationUtils.checkArgument(partialUpdateReaderContextOpt.isPresent(), "Partial update cannot be supported without PartialUpdateReaderContext");
+      partialUpdateReaderContextOpt.get().prepareReaderContext();
+    }
+
+    Option<PartialUpdateReaderContext> finalPartialUpdateReaderContextOpt = partialUpdateReaderContextOpt;
+    return operationPartialUpdatePair.map((SerializableFunction<Pair<CompactionOperation, Boolean>, List<WriteStatus>>) v1 -> compact(
```

Comment:
This will trigger the log reading again in `.mapToPair(operation -> Pair.of(operation, containsPartialUpdate(compactionHandler, metaClient, config, operation, compactionInstantTime, maxInstantTime,
            instantRange, taskContextSupplier, executionHelper)))`.  It's better to collect `operationPartialUpdatePair` to driver and parallelize it again to avoid repeated log reading.

Reply:
Can we just enable the fg reader compaction for spark then, so that the detection of partial update log files can be eliminated and the logic can be simplified a lot.